### PR TITLE
Add aldec-activehdl support

### PR DESCRIPTION
**Pull request by pstrueb**
With this proposed pull request support for the  aldec-activehdl simulator with SIM=activehdl is added. 

I Also added an extra_cmd parameter to the Simulator class. This is needed because I have use cases where I want to induce additional simulator command just before the actual run command. This is e.g. needed to be able to select certain signals or region of signals to be traced.
The extra_cmd feature could also be added for other simulators just before the run command. 

So far I was able to successfully test the the functionality with vhdl only because I have no license for simulating verilog.

### CODE REVIEW

**Comment by themperek** [comment]
So in cocotb there is `SCRIPT_FILE` but it is in a different place.
https://github.com/cocotb/cocotb/blob/8e8a7cce1123d213ec1485934e9ae53c0e8f027c/cocotb/share/makefiles/simulators/Makefile.questa#L137-L138

Not sure if the best way is to add more parameters since there seems no end to do this. 

Maybe we can modify this so one can extend it this way and make it much more flexible?:
``` python3

class ActivehdlCustom(Activehdl):
    def __init__(self, param, *argv, **kwargs):
        self.param = param
        super().__init__(*argv, **kwargs)

    def buid_script(self): # would add this to Activehdl class, could also spit into 2-3 functions 
        script = super().buid_script()
        script += self.param
        return script
```

Too complicated? Any opinion?

**Comment by pstrueb** [comment]
I dont think I fully understand the solution you are proposing. Are you saying that in the actual project we should inherit from the Simulator class we are using? Or are you proposing to add the sample code you provided directly to the corresponding simulator class (e.g. Activehdl) or make it part of class Simulator ?

Maybe you can describe the Idea a bit further, I would be happy the work further on this PR.

**Comment by themperek** [comment]
> Are you saying that in the actual project we should inherit from the Simulator class we are using?

Yes. This would be my proposal another way we will finish with a large number of arguments. This is some "more advance use" case.
We can split `build_command` into smaller functions. Like `build_script` or even `build_script_compile`, `build_script_elaborate`, ... so things are bit of issues to insert the code in custom class?

Or do we just add more extra parameters? But this easily may become ugly?

**Comment by pstrueb** [comment]
Okay I think than I understand the approach. On one hand it would be really flexible to implement "custom" functionality like this only in the verification project. On the other hand we would also have to make a re implementation for each simulator that we want to support which would basically lead to a re implementation of the simulator.py::run function as I unterstand it.

Another Argument for adding more extra parameters (e.g. extra_cmds) is that is would be an analogue impelentation to cocotbs SCRIPT_FILE which basically seems also to add extra commands (just stored in a script) in a defined location before the run command is executed.

So in short I would argument it would be cocotb fashioned to add more extra parameters

**Comment by themperek** [comment]
We can add `SCRIPT_FILE` parameter but in cocotb Makefiles it is in a different place than you have put `extra_cmd`. Is it important?

**Comment by pstrueb** [comment]
> We can add `SCRIPT_FILE` parameter but in cocotb Makefiles it is in a different place than you have put `extra_cmd`. Is it important?

Actually yes it is important - I tested it. The usecase I have is inserting trace commands for certain signals. This can only be done after the asim command initialized the Simulation. To be honest I didnt thought it would really matter.  So this is actually something a bit different than the SCRIPT_FILE parameter. Now I can better understand the argumentation that there seems no end to adding extra parameters, you are right. On the other hand it is a really easy solution with very few overhead.

**Comment by themperek** [comment]
>Now I can better understand the argumentation that there seems no end to adding extra parameters, you are right.

Maybe better to make it very easy to extent/change this class? Maye one can also use Mock.

**Comment by pstrueb** [comment]
With this commit we split the build_command method in the Activehdl class in smaller sub methods and wrap them. This way it found it convenient to alter the class methods the way it is needed in a verification project. For the wrapper we can use this code:

```
import os
from cocotb_test.simulator import Activehdl, Ghdl


class ActivehdlCustom(Activehdl):
    def __init__(self, extra_cmd=None, *argv, **kwargs):
        if extra_cmd is None:
            extra_cmd = []
        else:
            self.extra_cmd = extra_cmd
        super().__init__(*argv, **kwargs)

    def build_command_sim(self):
        script = super().build_command_sim()
        script += "\n".join(self.extra_cmd) + "\n"
        return script


def run(**kwargs):

    sim_env = os.environ["SIM"]

    if sim_env == "ghdl":
        sim = Ghdl(**kwargs)

    elif sim_env == "activehdl":
        sim = ActivehdlCustom(**kwargs)

    else:
        raise NotImplementedError("Simulator not wrapped yet for the project")

    return sim.run()
```

Then in the pytest testcases we import the custom run function and call it.

File: cocotb_test/simulator.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def __init__(
         compile_only=False,
         waves=None,
         gui=False,
-
+        extra_cmd=None,
```
Comment (themperek) [inline cocotb_test/simulator.py:59]:
```suggestion

```

File: cocotb_test/simulator.py [hunk 2]
```diff
@@ -769,12 +769,9 @@ def get_parameter_commands(self, parameters):
...
 
-    def build_command(self):
-
-        self.rtl_library = self.toplevel
-
-        do_script = "\nonerror {\n quit -code 1 \n} \n"
-
+    def build_command_compile(self):
+        do_script = ""
+
```
Comment (themperek) [inline cocotb_test/simulator.py:774]:
```suggestion
    def build_command(self):

        self.rtl_library = self.toplevel

        do_script = "\nonerror {\n quit -code 1 \n} \n"

```

**Review by themperek — COMMENTED** [review]
it looks like there are some leftovers.

File: cocotb_test/simulator.py [hunk 3]
```diff
@@ -832,6 +831,110 @@ def build_command(self):
...
+    def get_parameter_commands(self, parameters):
+        parameters_cmd = []
+        for name, value in parameters.items():
+            parameters_cmd.append("-g" + name + "=" + str(value))
+
+        return parameters_cmd
+
+
+    def build_command_compile(self):
```
Comment (themperek) [inline cocotb_test/simulator.py:857]: Is OK but could we call this `build_script_sim`?

Comment (pstrueb) [reply cocotb_test/simulator.py]: You mean like in this  ?

**Review by themperek — APPROVED** [review]

**Comment by themperek** [comment]
Thank you !
